db/db.go

```diff
@@ -245,9 +239,289 @@ func (db *DB) execImpl(basis types.Ref, function string, args types.List) (newDa
 	return newData, newDataChecksum, output, isWrite, nil
 }
 
+func removeTransaction(tx ReadTransaction) {
+	transactionMutex.Lock()
+	defer transactionMutex.Unlock()
+	delete(transactions, tx.ID())
+}
+
+func (db *DB) BeginReadTransaction() ReadTransaction {
+	transactionMutex.Lock()
+	defer transactionMutex.Unlock()
+	txID := transactionCounter
+	transactionCounter++
+	tx := &readTransaction{txID, db.noms, db.head}
+	transactions[txID] = tx
+	return tx
+}
+
+func (db *DB) ReadTransaction(body func(tx ReadTransaction)) error {
+	tx := db.BeginReadTransaction()
+	body(tx)
+	return tx.Close()
+}
+
+func (db *DB) CloseTransaction(txID int) error {
+	tx, err := db.FindReadTransaction(txID)
+	if err != nil {
+		return err
+	}
+	return tx.Close()
+}
+
+func (db *DB) CommitTransaction(txID int) (types.Ref, error) {
+	tx, err := db.FindWriteTransaction(txID)
+	if err != nil {
+		return types.Ref{}, err
+	}
+	return tx.Commit()
+}
+
+func (db *DB) AbortTransaction(txID int) error {
+	tx, err := db.FindWriteTransaction(txID)
+	if err != nil {
+		return err
+	}
+	return tx.Abort()
+}
+
+func (db *DB) BeginWriteTransaction() WriteTransaction {
+	transactionMutex.Lock()
+	defer transactionMutex.Unlock()
+	txID := transactionCounter
+	transactionCounter++
+	tx := &writeTransaction{txID, db, db.noms, db.head, db.head.Data(db.noms).Edit(), false}
+	transactions[txID] = tx
+	return tx
+}
+
+func (db *DB) WriteTransaction(body func(tx WriteTransaction)) error {
+	tx := db.BeginWriteTransaction()
+	body(tx)
+	_, err := tx.Commit()
+	return err
+}
+
 func (db *DB) lock() func() {
 	db.mu.Lock()
 	return func() {
 		db.mu.Unlock()
 	}
 }
+
+// ReadTransaction describes the operations allowed in a read transaction.
+type ReadTransaction interface {
```

Comment:
It's non-idiomatic to use interfaces this way in Go. Typically interfaces are at the caller of a boundary, not the callee.

Instead, prefer to expose the `readTransaction` struct (which might just become `transaction` -  see other comments).